src/mac.cpp

```diff
@@ -119,4 +124,98 @@ void HMAC::verify(const std::vector<uint8_t> &hmacValue)
     _impl->verify(hmacValue);
 }
 
+
+/* CMAC */
+
+class CMAC::Impl
+{
+public:
+    Impl(openssl::CmacCipherTypes cipherType, const std::vector<uint8_t> &key)
+    {
+        const EVP_CIPHER* cipher = openssl::_getCipherPtrFromCmacCipherType(cipherType);
+
+        if (key.empty()) {
+            throw MoCOCrWException("Key for CMAC is empty.");
+        }
+
+        _ctx = openssl::_CMAC_CTX_new();
+        openssl::_CMAC_Init(_ctx.get(), key, cipher, nullptr);
+    }
+
+    ~Impl() = default;
+
+    Impl(Impl&&) = default;
+
+    void update(const std::vector<uint8_t> &message)
+    {
+        if (_isFinished) {
+            throw MoCOCrWException("update() can't be called after finish()");
+        }
+        openssl::_CMAC_Update(_ctx.get(), message);
+    }
+
+    std::vector<uint8_t> finish()
+    {
+        if (_isFinished) {
+            throw MoCOCrWException("finish() can't be called twice.");
+        }
+
+        _result = openssl::_CMAC_Final(_ctx.get());
+
+        _isFinished = true;
+
+        return _result;
+    }
+
+    void verify(const std::vector<uint8_t> &cmacValue)
+    {
+        if (!_isFinished) {
+            finish();
+        }
+
+        if (cmacValue.size() != _result.size()) {
+            throw MoCOCrWException("CMAC verification failed. Length differs.");
+        }
+
+        if (CRYPTO_memcmp(cmacValue.data(), _result.data(), cmacValue.size())) {
+            throw MoCOCrWException("CMAC verification failed. Calculated value: " + utility::toHex(_result)
+                                   + ". Received value: " + utility::toHex(cmacValue));
+        }
+    }
+
+private:
+    openssl::SSL_CMAC_CTX_SharedPtr _ctx = nullptr;
```

Comment:
No hard rationale. It is done the same within the HMAC implementation and, since raw pointers are used for the next level of abstraction (within `CMAC::Impl`), I thought a `shared_ptr` reflect this fact the best.

Btw., my first intention was to extract the raw pointer as late as possible just for the calls of the `SSL_*` methods (i.e. within call of `*::callChecked()` in `openssl_wrap.cpp`), but I decided against it because it is used differently all over the place in the existing code. (I am not sure yet how much we are allowed to sacrifice consistency in favour of cleaner code.)

Reply:
No hard rationale. It is done the same within the HMAC implementation and, since raw pointers are used for the next level of abstraction (within `CMAC::Impl`), I thought a `shared_ptr` reflect this fact the best.

Btw., my first intention was to extract the raw pointer as late as possible just for the calls of the `SSL_*` methods (i.e. within call of `*::callChecked()` in `openssl_wrap.cpp`), but I decided against it because it is used differently all over the place in the existing code. (I am not sure yet how much we are allowed to sacrifice consistency in favour of cleaner code.)

Reply:
Then, I'd suggest to move to unique_ptr here. This is no public interface. We even change it to unique_ptr in the hmac without impacting anyone due to the impl pattern. Can you just change it in both places?

Reply:
Done.